File: C++/5_Longest_Palindromic_Substring.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        int i;
        string substring, sub = "";
        for (i = 0; i < s.size(); i++) {
            substring = ifPalin(s, i, i);
            if (substring.size() > sub.size())
                sub = substring;
            substring = ifPalin(s, i, i + 1);
            if (substring.size() > sub.size())
                sub = substring;
        }
        return sub;
    }

    string ifPalin(string s, int start, int end) {
        while (start >= 0 and end < s.size() and s[start] == s[end]) {
            start -= 1;
            end += 1;
        }
        return s.substr(start + 1, end - start - 1);
    }
};


int main() {
    Solution s;
    cout << s.longestPalindrome("abb") << endl;
    cout << s.longestPalindrome("aba") << endl;


}

File: python/5_Longest_Palindromic_Substring.py
class Solution(object):
    def longestPalindrome(self, s):
        substring = ''
        for i in xrange(len(s)):
            subS = self.ifPalin(s, i, i)
            if len(subS) > len(substring):
                substring = subS
            # for even case
            subS = self.ifPalin(s, i, i + 1)
            if len(subS) > len(substring):
                substring = subS
            
        return substring

    def ifPalin(self, s, start, end):
        while start >= 0 and end < len(s) and s[start] == s[end]:
            start -= 1
            end += 1

        return s[start + 1:end]